File: tools/tf-types-ts/src/core/bridge-tls.ts
/**
 * TLS / mTLS bridge — accept a peer-supplied X.509 certificate chain,
 * verify it against a configured set of trust anchors, and project the
 * verified leaf into a TrustForge actor identity + capabilities.
 *
 * Uses Bun's built-in `node:crypto` X509Certificate API (a libuv-backed
 * binding to BoringSSL) for parsing and signature verification — we do
 * not roll our own ASN.1 path validator.
 */

import { X509Certificate, createPublicKey } from "node:crypto";

import type { ActorIdentity } from "../generated/actor-identity.js";
import type { Capability } from "../generated/_common.js";
import { BridgeFailure, type Bridge, type BridgeKind } from "./bridges.js";

export interface TlsBridgeConfig {
  bridgeId: string;
  /** Trust domain that owns this bridge instance. */
  trustDomain: string;
  /** PEM-encoded trust anchors. The leaf chain must terminate in one of these. */
  rootCertificatesPem: string[];
  /** Optional clock override for testing — defaults to `Date.now()`. */
  now?: () => Date;
  /** Maximum acceptable chain length (leaf + intermediates + root). Default 6. */
  maxChainLength?: number;
  /** Optional SAN URI predicate. When set, the leaf must list a URI SAN
   *  matching this value (typical pattern: pin to a SPIFFE ID). */
  requiredSanUri?: string;
  /** Optional override that turns each leaf EKU OID into a TrustForge action.
   *  Default: use the table in `EKU_TO_ACTION`. */
  ekuToAction?: (oid: string) => string | undefined;
}

export interface TlsVerificationResult {
  identity: ActorIdentity;
  capabilities: Capability[];
  leaf: X509Certificate;
  chain: X509Certificate[];
}

/** Default mapping from X.509 Extended Key Usage OIDs to TrustForge actions. */
export const EKU_TO_ACTION: Record<string, string> = {
  "1.3.6.1.5.5.7.3.1": "tls.server-auth",
  "1.3.6.1.5.5.7.3.2": "tls.client-auth",
  "1.3.6.1.5.5.7.3.3": "code.sign",
  "1.3.6.1.5.5.7.3.4": "email.protect",
  "1.3.6.1.5.5.7.3.8": "timestamp.sign",
  "1.3.6.1.5.5.7.3.9": "ocsp.sign",
};

/** Parse one or more PEM blocks (whitespace-separated) into X509 cert objects. */
export function parsePemBundle(pem: string): X509Certificate[] {
  const blocks = pem.match(/-----BEGIN CERTIFICATE-----[\s\S]*?-----END CERTIFICATE-----/g);
  if (!blocks || blocks.length === 0) {
    throw new BridgeFailure({
      code: "invalid-input",
      message: "PEM bundle contains no CERTIFICATE blocks",
    });
  }
  return blocks.map((b) => new X509Certificate(b));
}

export class TlsBridge implements Bridge {
  readonly kind: BridgeKind = "tls";
  readonly bridgeId: string;
  readonly trustDomain: string;
  private readonly cfg: TlsBridgeConfig;
  private readonly roots: X509Certificate[];

  constructor(cfg: TlsBridgeConfig) {
    this.bridgeId = cfg.bridgeId;
    this.trustDomain = cfg.trustDomain;
    this.cfg = cfg;
    if (!cfg.rootCertificatesPem || cfg.rootCertificatesPem.length === 0) {
      throw new BridgeFailure({
        code: "invalid-input",
        message: "TLS bridge requires at least one trust anchor",
      });
    }
    this.roots = cfg.rootCertificatesPem.map((p, i) => {
      try {
        return parsePemBundle(p)[0]!;
      } catch (e) {
        throw new BridgeFailure({
          code: "invalid-input",
          message: `root[${i}] failed to parse: ${(e as Error).message}`,
        });
      }
    });
  }

  /** Verify a peer chain (leaf first, then intermediates) and return the
   *  projected TrustForge identity + capabilities derived from the leaf. */
  verifyChain(chainPem: string[] | string): TlsVerificationResult {
    const certs =
      typeof chainPem === "string"
        ? parsePemBundle(chainPem)
        : chainPem.flatMap((p) => parsePemBundle(p));
    if (certs.length === 0) {
      throw new BridgeFailure({ code: "invalid-input", message: "empty chain" });
    }
    const max = this.cfg.maxChainLength ?? 6;
    if (certs.length > max) {
      throw new BridgeFailure({
        code: "rejected",
        message: `chain longer than max (${certs.length} > ${max})`,
      });
    }
    const now = (this.cfg.now ?? (() => new Date()))();

    // Validate every cert's validity window first — cheap and avoids verifying
    // signatures on already-expired material.
    for (const c of certs) {
      const notBefore = new Date(c.validFrom);
      const notAfter = new Date(c.validTo);
      if (Number.isNaN(notBefore.getTime()) || Number.isNaN(notAfter.getTime())) {
        throw new BridgeFailure({
          code: "rejected",
          message: `cert ${c.subject} has unparseable validity window`,
        });
      }
      if (now < notBefore) {
        throw new BridgeFailure({
          code: "rejected",
          message: `cert ${c.subject} not yet valid (notBefore=${notBefore.toISOString()})`,
        });
      }
      if (now > notAfter) {
        throw new BridgeFailure({
          code: "rejected",
          message: `cert ${c.subject} expired (notAfter=${notAfter.toISOString()})`,
        });
      }
    }

    // Walk leaf → intermediates verifying each link, then verify the final
    // intermediate against a configured root.
    const leaf = certs[0]!;
    const chain: X509Certificate[] = [leaf];
    let current = leaf;
    for (let depth = 0; depth < max; depth++) {
      const issuer =
        certs.slice(1).find((c) => c.subject === current.issuer && c !== current) ??
        this.roots.find((r) => r.subject === current.issuer);
      if (!issuer) {
        throw new BridgeFailure({
          code: "rejected",
          message: `no issuer cert for ${current.subject} (issuer=${current.issuer})`,
        });
      }
      // BoringSSL verifies the cert's signature using the issuer public key;
      // returns false on mismatch, throws on malformed input.
      const ok = current.verify(issuer.publicKey);
      if (!ok) {
        throw new BridgeFailure({
          code: "rejected",
          message: `signature verification failed for ${current.subject}`,
        });
      }
      if (this.roots.includes(issuer)) {
        chain.push(issuer);
        // Confirm the root is self-signed and within validity.
        const selfOk = issuer.verify(issuer.publicKey);
        if (!selfOk) {
          throw new BridgeFailure({
            code: "rejected",
            message: `root ${issuer.subject} not self-consistent`,
          });
        }
        return this.project(leaf, chain);
      }
      chain.push(issuer);
      current = issuer;
    }
    throw new BridgeFailure({
      code: "rejected",
      message: `chain exceeds max depth ${max} without reaching a trust anchor`,
    });
  }

  private project(leaf: X509Certificate, chain: X509Certificate[]): TlsVerificationResult {
    if (this.cfg.requiredSanUri) {
      const sans = parseSanUris(leaf);
      if (!sans.includes(this.cfg.requiredSanUri)) {
        throw new BridgeFailure({
          code: "rejected",
          message: `leaf SAN URIs ${JSON.stringify(sans)} missing required ${this.cfg.requiredSanUri}`,
        });
      }
    }
    const sanUris = parseSanUris(leaf);
    const sanDns = parseSanDns(leaf);
    // Prefer a spiffe:// URI SAN as the actor binding; else CN; else first DNS SAN.
    const spiffeSan = sanUris.find((u) => u.startsWith("spiffe://"));
    const cn = parseCommonName(leaf.subject);
    const subject = spiffeSan ?? cn ?? sanDns[0] ?? leaf.subject;
    const actorType: ActorIdentity["actor_type"] = spiffeSan ? "service" : "device";
    const actorId = `tf:actor:${actorType}:${this.cfg.trustDomain}/${encodeActorPath(subject)}`;
    const publicKey = leaf.publicKey;
    const jwk = publicKey.export({ format: "jwk" }) as Record<string, unknown>;
    const algorithm = jwkAlgorithmName(jwk);
    const publicKeyB64 = encodePublicKey(jwk);

    const identity: ActorIdentity = {
      identity_version: "1",
      actor_id: actorId,
      actor_type: actorType,
      public_keys: [
        {
          key_id: leaf.fingerprint256.replace(/:/g, "").toLowerCase(),
          algorithm,
          public_key: publicKeyB64,
          purpose: "signing",
        },
      ],
      trust_levels: [actorType === "service" ? "T4" : "T3"],
      authority_roots: chain.slice(-1).map((root) => ({
        kind: "organization" as const,
        id: parseCommonName(root.subject) ?? root.subject,
      })),
      valid_from: new Date(leaf.validFrom).toISOString(),
      valid_until: new Date(leaf.validTo).toISOString(),
    };

    const capabilities = ekusForLeaf(leaf).flatMap((oid) => {
      const action = (this.cfg.ekuToAction ?? ((o) => EKU_TO_ACTION[o]))(oid);
      return action ? [{ name: action, risk: "R2" as const }] : [];
    });

    return { identity, capabilities, leaf, chain };
  }
}

/** Pull URI SANs out of a certificate (best-effort, RFC 5280 §4.2.1.6). */
export function parseSanUris(cert: X509Certificate): string[] {
  const san = (cert as unknown as { subjectAltName?: string }).subjectAltName;
  if (!san) return [];
  const out: string[] = [];
  for (const part of san.split(",").map((s) => s.trim())) {
    const m = /^URI:(.*)$/.exec(part);
    if (m) out.push(m[1]!);
  }
  return out;
}

export function parseSanDns(cert: X509Certificate): string[] {
  const san = (cert as unknown as { subjectAltName?: string }).subjectAltName;
  if (!san) return [];
  const out: string[] = [];
  for (const part of san.split(",").map((s) => s.trim())) {
    const m = /^DNS:(.*)$/.exec(part);
    if (m) out.push(m[1]!);
  }
  return out;
}

function parseCommonName(distinguishedName: string): string | undefined {
  for (const part of distinguishedName.split(/[,\n]/).map((s) => s.trim())) {
    const m = /^CN\s*=\s*(.+)$/i.exec(part);
    if (m) return m[1]!;
  }
  return undefined;
}

function encodeActorPath(s: string): string {
  let out = "";
  for (const ch of s) {
    if (/[A-Za-z0-9_.~/-]/.test(ch)) out += ch;
    else out += encodeURIComponent(ch);
  }
  return out;
}

function jwkAlgorithmName(jwk: Record<string, unknown>): "ed25519" | "p256" | "p384" | "rsa" {
  const kty = jwk["kty"];
  const crv = jwk["crv"];
  if (kty === "OKP" && crv === "Ed25519") return "ed25519";
  if (kty === "EC" && crv === "P-256") return "p256";
  if (kty === "EC" && crv === "P-384") return "p384";
  if (kty === "RSA") return "rsa";
  throw new BridgeFailure({
    code: "unsupported",
    message: `unsupported JWK kty/crv: ${kty}/${crv}`,
  });
}

function encodePublicKey(jwk: Record<string, unknown>): string {
  // For Ed25519 we ship the raw 32-byte x. For EC P-256/P-384 we ship
  // uncompressed SEC1 (0x04||x||y). For RSA we ship the SubjectPublicKeyInfo
  // DER (re-derived via createPublicKey). All base64-encoded.
  const kty = jwk["kty"];
  if (kty === "OKP") {
    return base64FromBase64Url(String(jwk["x"]));
  }
  if (kty === "EC") {
    const x = base64UrlToBytes(String(jwk["x"]));
    const y = base64UrlToBytes(String(jwk["y"]));
    const buf = new Uint8Array(1 + x.length + y.length);
    buf[0] = 0x04;
    buf.set(x, 1);
    buf.set(y, 1 + x.length);
    return Buffer.from(buf).toString("base64");
  }
  if (kty === "RSA") {
    const k = createPublicKey({ key: jwk as object, format: "jwk" });
    return k.export({ format: "der", type: "spki" }).toString("base64");
  }
  throw new BridgeFailure({ code: "unsupported", message: `unsupported jwk kty: ${kty}` });
}

function base64FromBase64Url(b64u: string): string {
  return Buffer.from(base64UrlToBytes(b64u)).toString("base64");
}

function base64UrlToBytes(b64u: string): Uint8Array {
  let s = b64u.replace(/-/g, "+").replace(/_/g, "/");
  while (s.length % 4 !== 0) s += "=";
  return new Uint8Array(Buffer.from(s, "base64"));
}

function ekusForLeaf(cert: X509Certificate): string[] {
  // node:crypto's X509Certificate doesn't expose EKUs directly, but the
  // certificate's `infoAccess` covers AIA OIDs. For EKU we parse the raw DER
  // ASN.1 of the extension: id-ce-extKeyUsage OID is 2.5.29.37.
  const der = cert.raw;
  return extractEkuOids(new Uint8Array(der));
}

/** Lightweight DER walker: find the extKeyUsage extension OID 2.5.29.37 and
 *  return the list of OIDs inside its SEQUENCE OF OBJECT IDENTIFIER. We
 *  intentionally avoid pulling in a full ASN.1 library here. */
export function extractEkuOids(der: Uint8Array): string[] {
  const eku = findExtension(der, "2.5.29.37");
  if (!eku) return [];
  // EKU value is OCTET STRING wrapping a SEQUENCE OF OID.
  return readOidSequence(eku);
}

function findExtension(der: Uint8Array, targetOid: string): Uint8Array | null {
  // Walk: top SEQUENCE = Certificate; first child = TBSCertificate; we descend
  // until we find a SEQUENCE OF Extension blocks.
  const cert = readSequence(der, 0);
  if (!cert) return null;
  const tbs = readSequence(der, cert.contentStart);
  if (!tbs) return null;
  // Inside TBSCertificate, walk fields until we hit the [3] extensions tag.
  let pos = tbs.contentStart;
  const end = tbs.contentStart + tbs.contentLength;
  while (pos < end) {
    const tag = der[pos]!;
    const len = readLength(der, pos + 1);
    if (!len) return null;
    if (tag === 0xa3) {
      // [3] EXPLICIT Extensions
      const inner = readSequence(der, pos + 1 + len.headerSize);
      if (!inner) return null;
      let p = inner.contentStart;
      const innerEnd = inner.contentStart + inner.contentLength;
      while (p < innerEnd) {
        const ext = readSequence(der, p);
        if (!ext) return null;
        const oid = readOid(der, ext.contentStart);
        if (!oid) return null;
        if (oid.value === targetOid) {
          // Skip optional BOOLEAN critical, then OCTET STRING containing extn value.
          let q = oid.nextPos;
          if (der[q] === 0x01) {
            const bl = readLength(der, q + 1);
            if (!bl) return null;
            q = q + 1 + bl.headerSize + bl.length;
          }
          if (der[q] !== 0x04) return null;
          const ol = readLength(der, q + 1);
          if (!ol) return null;
          return der.subarray(q + 1 + ol.headerSize, q + 1 + ol.headerSize + ol.length);
        }
        p = ext.contentStart + ext.contentLength;
      }
      return null;
    }
    pos = pos + 1 + len.headerSize + len.length;
  }
  return null;
}

function readOidSequence(extnValue: Uint8Array): string[] {
  const seq = readSequence(extnValue, 0);
  if (!seq) return [];
  const out: string[] = [];
  let p = seq.contentStart;
  const end = seq.contentStart + seq.contentLength;
  while (p < end) {
    const oid = readOid(extnValue, p);
    if (!oid) break;
    out.push(oid.value);
    p = oid.nextPos;
  }
  return out;
}

interface SeqHeader {
  contentStart: number;
  contentLength: number;
}

function readSequence(buf: Uint8Array, pos: number): SeqHeader | null {
  if (buf[pos] !== 0x30) return null;
  const len = readLength(buf, pos + 1);
  if (!len) return null;
  return { contentStart: pos + 1 + len.headerSize, contentLength: len.length };
}

function readLength(
  buf: Uint8Array,
  pos: number,
): { length: number; headerSize: number } | null {
  if (pos >= buf.length) return null;
  const b = buf[pos]!;
  if (b < 0x80) return { length: b, headerSize: 1 };
  const n = b & 0x7f;
  if (n === 0 || n > 4 || pos + n >= buf.length) return null;
  let len = 0;
  for (let i = 0; i < n; i++) len = (len << 8) | buf[pos + 1 + i]!;
  return { length: len, headerSize: 1 + n };
}

function readOid(buf: Uint8Array, pos: number): { value: string; nextPos: number } | null {
  if (buf[pos] !== 0x06) return null;
  const len = readLength(buf, pos + 1);
  if (!len) return null;
  const start = pos + 1 + len.headerSize;
  const data = buf.subarray(start, start + len.length);
  if (data.length === 0) return null;
  const first = data[0]!;
  const parts: number[] = [Math.floor(first / 40), first % 40];
  let acc = 0;
  for (let i = 1; i < data.length; i++) {
    acc = (acc << 7) | (data[i]! & 0x7f);
    if ((data[i]! & 0x80) === 0) {
      parts.push(acc);
      acc = 0;
    }
  }
  return { value: parts.join("."), nextPos: start + len.length };
}

File: crates/tf-types/src/bridge_tls.rs
//! TLS / mTLS bridge — accept a peer-supplied X.509 certificate chain,
//! verify it against a configured set of trust anchors, and project the
//! verified leaf into a TrustForge actor identity + capabilities.
//!
//! Uses `x509-parser` for ASN.1 parsing and signature verification, so we
//! avoid embedding our own ASN.1/DER walker.

use std::collections::HashSet;
use std::time::{SystemTime, UNIX_EPOCH};

use x509_parser::certificate::X509Certificate;
use x509_parser::extensions::{GeneralName, ParsedExtension};
use x509_parser::pem::Pem;
use x509_parser::prelude::FromDer;
use x509_parser::time::ASN1Time;

use crate::bridges::{Bridge, BridgeError, BridgeKind};
use crate::generated::{
    ActorIdentity, ActorIdentity_IdentityVersion, ActorType, AuthorityRoot, AuthorityRoot_Kind,
    PublicKey, PublicKey_Purpose, TrustLevel,
};

/// Mapping from X.509 Extended Key Usage OIDs to TrustForge action names.
pub fn default_eku_to_action(oid: &str) -> Option<&'static str> {
    match oid {
        "1.3.6.1.5.5.7.3.1" => Some("tls.server-auth"),
        "1.3.6.1.5.5.7.3.2" => Some("tls.client-auth"),
        "1.3.6.1.5.5.7.3.3" => Some("code.sign"),
        "1.3.6.1.5.5.7.3.4" => Some("email.protect"),
        "1.3.6.1.5.5.7.3.8" => Some("timestamp.sign"),
        "1.3.6.1.5.5.7.3.9" => Some("ocsp.sign"),
        _ => None,
    }
}

#[derive(Clone, Debug)]
pub struct TlsBridgeConfig {
    pub bridge_id: String,
    pub trust_domain: String,
    pub root_certificates_pem: Vec<String>,
    pub max_chain_length: Option<usize>,
    pub required_san_uri: Option<String>,
    pub now_unix_seconds: Option<u64>,
}

#[derive(Clone, Debug)]
pub struct TlsVerificationResult {
    pub identity: ActorIdentity,
    pub capabilities: Vec<String>,
    pub leaf_subject: String,
    pub chain_subjects: Vec<String>,
}

pub struct TlsBridge {
    cfg: TlsBridgeConfig,
    roots: Vec<Vec<u8>>, // owned DER for each configured root
}

impl TlsBridge {
    pub fn new(cfg: TlsBridgeConfig) -> Result<Self, BridgeError> {
        if cfg.root_certificates_pem.is_empty() {
            return Err(BridgeError::InvalidInput(
                "TLS bridge requires at least one trust anchor".into(),
            ));
        }
        let mut roots = Vec::with_capacity(cfg.root_certificates_pem.len());
        for (i, pem) in cfg.root_certificates_pem.iter().enumerate() {
            let der = parse_single_pem(pem)
                .map_err(|e| BridgeError::InvalidInput(format!("root[{}]: {}", i, e)))?;
            roots.push(der);
        }
        Ok(TlsBridge { cfg, roots })
    }

    pub fn verify_chain(&self, chain_pem: &[String]) -> Result<TlsVerificationResult, BridgeError> {
        let mut chain_der: Vec<Vec<u8>> = Vec::new();
        for (i, pem) in chain_pem.iter().enumerate() {
            for der in parse_pem_bundle(pem)
                .map_err(|e| BridgeError::InvalidInput(format!("chain[{}]: {}", i, e)))?
            {
                chain_der.push(der);
            }
        }
        if chain_der.is_empty() {
            return Err(BridgeError::InvalidInput("empty chain".into()));
        }
        let max = self.cfg.max_chain_length.unwrap_or(6);
        if chain_der.len() > max {
            return Err(BridgeError::Rejected(format!(
                "chain longer than max ({} > {})",
                chain_der.len(),
                max
            )));
        }

        let now = self
            .cfg
            .now_unix_seconds
            .unwrap_or_else(|| SystemTime::now().duration_since(UNIX_EPOCH).unwrap().as_secs());
        let now_asn1 = ASN1Time::from_timestamp(now as i64)
            .map_err(|e| BridgeError::InvalidInput(format!("now overflow: {}", e)))?;

        // Parse all certs once; we'll re-borrow as we walk.
        let parsed: Vec<X509Certificate> = chain_der
            .iter()
            .map(|d| {
                let (_, c) = X509Certificate::from_der(d)
                    .map_err(|e| BridgeError::InvalidInput(format!("DER parse: {}", e)))?;
                Ok::<_, BridgeError>(c)
            })
            .collect::<Result<_, _>>()?;

        // Validity windows for everything in the chain.
        for c in &parsed {
            let validity = c.validity();
            if !validity.is_valid_at(now_asn1) {
                return Err(BridgeError::Rejected(format!(
                    "cert {} outside validity window",
                    c.subject()
                )));
            }
        }

        // Walk leaf → chain → root.
        let leaf = &parsed[0];
        let mut chain_subjects: Vec<String> = vec![leaf.subject().to_string()];
        let mut current_idx: Option<usize> = Some(0);
        let mut visited: HashSet<String> = HashSet::new();
        visited.insert(leaf.subject().to_string());

        let roots_parsed: Vec<X509Certificate> = self
            .roots
            .iter()
            .map(|d| X509Certificate::from_der(d).map(|p| p.1))
            .collect::<Result<_, _>>()
            .map_err(|e| BridgeError::Internal(format!("root DER reparse: {}", e)))?;

        for _ in 0..max {
            let cur = &parsed[current_idx.expect("current set")];
            // Try to find an issuer in the supplied chain (other than `cur`).
            let inter_idx = parsed
                .iter()
                .enumerate()
                .find(|(i, c)| *i != current_idx.unwrap() && c.subject() == cur.issuer())
                .map(|(i, _)| i);
            let issuer_in_chain = inter_idx.map(|i| &parsed[i]);
            let root_match = roots_parsed.iter().find(|r| r.subject() == cur.issuer());

            let issuer = match issuer_in_chain.or(root_match) {
                Some(c) => c,
                None => {
                    return Err(BridgeError::Rejected(format!(
                        "no issuer cert for {} (issuer={})",
                        cur.subject(),
                        cur.issuer()
                    )))
                }
            };

            cur.verify_signature(Some(issuer.public_key()))
                .map_err(|e| {
                    BridgeError::Rejected(format!(
                        "signature verification failed for {}: {}",
                        cur.subject(),
                        e
                    ))
                })?;

            chain_subjects.push(issuer.subject().to_string());
            if root_match.is_some() && issuer_in_chain.is_none() {
                // Reached a configured trust anchor; validate the root is
                // self-signed and current.
                issuer
                    .verify_signature(Some(issuer.public_key()))
                    .map_err(|e| {
                        BridgeError::Rejected(format!(
                            "root {} not self-consistent: {}",
                            issuer.subject(),
                            e
                        ))
                    })?;
                return self.project(leaf, issuer, chain_subjects);
            }
            current_idx = inter_idx;
            if !visited.insert(issuer.subject().to_string()) {
                return Err(BridgeError::Rejected("chain loop detected".into()));
            }
        }
        Err(BridgeError::Rejected(format!(
            "chain exceeds max depth {} without reaching trust anchor",
            max
        )))
    }

    fn project(
        &self,
        leaf: &X509Certificate,
        root: &X509Certificate,
        chain_subjects: Vec<String>,
    ) -> Result<TlsVerificationResult, BridgeError> {
        let san_uris = collect_san_uris(leaf);
        if let Some(req) = &self.cfg.required_san_uri {
            if !san_uris.iter().any(|u| u == req) {
                return Err(BridgeError::Rejected(format!(
                    "leaf SAN URIs {:?} missing required {}",
                    san_uris, req
                )));
            }
        }
        let cn = parse_common_name(&leaf.subject().to_string());
        let san_dns = collect_san_dns(leaf);
        let spiffe_san = san_uris.iter().find(|u| u.starts_with("spiffe://")).cloned();
        let subject = spiffe_san
            .clone()
            .or(cn.clone())
            .or_else(|| san_dns.first().cloned())
            .unwrap_or_else(|| leaf.subject().to_string());
        let actor_type = if spiffe_san.is_some() {
            ActorType::Service
        } else {
            ActorType::Device
        };
        let type_str = match actor_type {
            ActorType::Service => "service",
            _ => "device",
        };
        let actor_id = format!(
            "tf:actor:{}:{}/{}",
            type_str,
            self.cfg.trust_domain,
            encode_actor_path(&subject)
        );

        let pk = leaf.public_key();
        let alg_oid = pk.algorithm.algorithm.to_id_string();
        let algorithm = match alg_oid.as_str() {
            "1.2.840.113549.1.1.1" => "rsa",
            "1.2.840.10045.2.1" => "p256",
            "1.3.101.112" => "ed25519",
            _ => "unknown",
        };
        let public_key_b64 = base64::engine::general_purpose::STANDARD
            .encode(pk.subject_public_key.data.as_ref());

        let fingerprint = sha256_hex(leaf.as_ref());

        let identity = ActorIdentity {
            identity_version: ActorIdentity_IdentityVersion::V1,
            actor_id,
            actor_type: actor_type.clone(),
            instance_id: None,
            public_keys: vec![PublicKey {
                key_id: fingerprint,
                algorithm: algorithm.to_string(),
                public_key: public_key_b64,
                purpose: PublicKey_Purpose::Signing,
                valid_from: None,
                valid_until: None,
            }],
            trust_levels: vec![if matches!(actor_type, ActorType::Service) {
                TrustLevel::T4
            } else {
                TrustLevel::T3
            }],
            authority_roots: vec![AuthorityRoot {
                kind: AuthorityRoot_Kind::Organization,
                id: parse_common_name(&root.subject().to_string())
                    .unwrap_or_else(|| root.subject().to_string()),
            }],
            attestations: None,
            valid_from: rfc3339_from_unix(leaf.validity().not_before.timestamp()),
            valid_until: Some(rfc3339_from_unix(leaf.validity().not_after.timestamp())),
            revocation_ref: None,
            signature: None,
        };

        let capabilities = collect_eku_actions(leaf);

        Ok(TlsVerificationResult {
            identity,
            capabilities,
            leaf_subject: leaf.subject().to_string(),
            chain_subjects,
        })
    }
}

impl Bridge for TlsBridge {
    fn bridge_id(&self) -> &str {
        &self.cfg.bridge_id
    }
    fn kind(&self) -> BridgeKind {
        BridgeKind::Tls
    }
    fn trust_domain(&self) -> &str {
        &self.cfg.trust_domain
    }
}

fn parse_single_pem(pem: &str) -> Result<Vec<u8>, String> {
    let mut all = parse_pem_bundle(pem)?;
    if all.is_empty() {
        return Err("no CERTIFICATE block".into());
    }
    Ok(all.remove(0))
}

fn parse_pem_bundle(pem: &str) -> Result<Vec<Vec<u8>>, String> {
    let mut bytes = pem.as_bytes();
    let mut out = Vec::new();
    while !bytes.is_empty() {
        match Pem::read(std::io::Cursor::new(bytes)) {
            Ok((p, consumed)) => {
                if p.label != "CERTIFICATE" {
                    bytes = &bytes[consumed..];
                    continue;
                }
                out.push(p.contents);
                bytes = &bytes[consumed..];
            }
            Err(_) => break,
        }
    }
    Ok(out)
}

fn collect_san_uris(cert: &X509Certificate) -> Vec<String> {
    cert.extensions()
        .iter()
        .flat_map(|ext| match ext.parsed_extension() {
            ParsedExtension::SubjectAlternativeName(san) => san
                .general_names
                .iter()
                .filter_map(|gn| match gn {
                    GeneralName::URI(u) => Some(u.to_string()),
                    _ => None,
                })
                .collect::<Vec<_>>(),
            _ => Vec::new(),
        })
        .collect()
}

fn collect_san_dns(cert: &X509Certificate) -> Vec<String> {
    cert.extensions()
        .iter()
        .flat_map(|ext| match ext.parsed_extension() {
            ParsedExtension::SubjectAlternativeName(san) => san
                .general_names
                .iter()
                .filter_map(|gn| match gn {
                    GeneralName::DNSName(d) => Some(d.to_string()),
                    _ => None,
                })
                .collect::<Vec<_>>(),
            _ => Vec::new(),
        })
        .collect()
}

fn collect_eku_actions(cert: &X509Certificate) -> Vec<String> {
    let mut out = Vec::new();
    for ext in cert.extensions() {
        if let ParsedExtension::ExtendedKeyUsage(eku) = ext.parsed_extension() {
            if eku.any {
                continue;
            }
            for oid in &eku.other {
                if let Some(action) = default_eku_to_action(&oid.to_id_string()) {
                    out.push(action.to_string());
                }
            }
            if eku.client_auth {
                out.push("tls.client-auth".to_string());
            }
            if eku.server_auth {
                out.push("tls.server-auth".to_string());
            }
            if eku.code_signing {
                out.push("code.sign".to_string());
            }
            if eku.email_protection {
                out.push("email.protect".to_string());
            }
            if eku.time_stamping {
                out.push("timestamp.sign".to_string());
            }
            if eku.ocsp_signing {
                out.push("ocsp.sign".to_string());
            }
        }
    }
    // Deduplicate while preserving order.
    let mut seen = HashSet::new();
    out.into_iter()
        .filter(|s| seen.insert(s.clone()))
        .collect()
}

fn parse_common_name(distinguished_name: &str) -> Option<String> {
    for part in distinguished_name.split(['\n', ',']) {
        let trimmed = part.trim();
        if let Some(rest) = trimmed
            .strip_prefix("CN=")
            .or_else(|| trimmed.strip_prefix("cn="))
        {
            return Some(rest.to_string());
        }
    }
    None
}

fn encode_actor_path(s: &str) -> String {
    let mut out = String::with_capacity(s.len());
    for b in s.bytes() {
        match b {
            b'A'..=b'Z' | b'a'..=b'z' | b'0'..=b'9' | b'-' | b'_' | b'.' | b'~' | b'/' => {
                out.push(b as char);
            }
            _ => out.push_str(&format!("%{:02X}", b)),
        }
    }
    out
}

fn sha256_hex(bytes: &[u8]) -> String {
    use sha2::{Digest, Sha256};
    let digest = Sha256::digest(bytes);
    digest.iter().map(|b| format!("{:02x}", b)).collect()
}

fn rfc3339_from_unix(secs: i64) -> String {
    let (y, m, d, h, mi, s) = secs_to_ymdhms(secs);
    format!("{:04}-{:02}-{:02}T{:02}:{:02}:{:02}Z", y, m, d, h, mi, s)
}

fn secs_to_ymdhms(secs: i64) -> (i32, u32, u32, u32, u32, u32) {
    let days = secs.div_euclid(86_400);
    let time = secs.rem_euclid(86_400);
    let hour = (time / 3600) as u32;
    let minute = ((time % 3600) / 60) as u32;
    let second = (time % 60) as u32;
    let z = days + 719_468;
    let era = if z >= 0 { z } else { z - 146_096 } / 146_097;
    let doe = (z - era * 146_097) as u64;
    let yoe = (doe - doe / 1460 + doe / 36524 - doe / 146096) / 365;
    let y = yoe as i64 + era * 400;
    let doy = doe - (365 * yoe + yoe / 4 - yoe / 100);
    let mp = (5 * doy + 2) / 153;
    let d = (doy - (153 * mp + 2) / 5 + 1) as u32;
    let m = if mp < 10 { (mp + 3) as u32 } else { (mp - 9) as u32 };
    let year = if m <= 2 { y + 1 } else { y };
    (year as i32, m, d, hour, minute, second)
}

use base64::Engine as _;
